crates/fluvio/src/producer/partition_producer.rs

```diff
@@ -0,0 +1,238 @@
+use std::sync::Arc;
+use std::collections::VecDeque;
+use std::time::Duration;
+
+use async_lock::{Mutex, RwLock};
+use dataplane::ReplicaKey;
+use dataplane::batch::Batch;
+use dataplane::produce::{DefaultPartitionRequest, DefaultTopicRequest, DefaultProduceRequest};
+use fluvio_future::timer::sleep;
+use fluvio_types::SpuId;
+use fluvio_types::event::StickyEvent;
+use tracing::{debug, info, instrument, error, trace};
+
+use crate::error::{Result, FluvioError};
+use crate::spu::SpuPool;
+
+use super::ProducerError;
+use super::accumulator::{ProducerBatch, BatchEvents};
+use super::event::EventHandler;
+
+/// Struct that is responsible for sending produce requests to the SPU in a given partition.
+pub(crate) struct PartitionProducer {
+    replica: ReplicaKey,
+    spu_pool: Arc<SpuPool>,
+    batches_lock: Arc<Mutex<VecDeque<ProducerBatch>>>,
+    batch_events: Arc<BatchEvents>,
+    linger: Duration,
+    last_error: Arc<RwLock<Option<ProducerError>>>,
+}
+
+impl PartitionProducer {
+    fn new(
+        replica: ReplicaKey,
+        spu_pool: Arc<SpuPool>,
+        batches_lock: Arc<Mutex<VecDeque<ProducerBatch>>>,
+        batch_events: Arc<BatchEvents>,
+        linger: Duration,
+        last_error: Arc<RwLock<Option<ProducerError>>>,
+    ) -> Self {
+        Self {
+            replica,
+            spu_pool,
+            batches_lock,
+            batch_events,
+            linger,
+            last_error,
+        }
+    }
+
+    pub fn shared(
+        replica: ReplicaKey,
+        spu_pool: Arc<SpuPool>,
+        batches: Arc<Mutex<VecDeque<ProducerBatch>>>,
+        batch_events: Arc<BatchEvents>,
+        linger: Duration,
+        error: Arc<RwLock<Option<ProducerError>>>,
+    ) -> Arc<Self> {
+        Arc::new(PartitionProducer::new(
+            replica,
+            spu_pool,
+            batches,
+            batch_events,
+            linger,
+            error,
+        ))
+    }
+
+    #[allow(clippy::too_many_arguments)]
+    pub(crate) fn start(
+        replica: ReplicaKey,
+        spu_pool: Arc<SpuPool>,
+        batches: Arc<Mutex<VecDeque<ProducerBatch>>>,
+        batch_events: Arc<BatchEvents>,
+        linger: Duration,
+        error: Arc<RwLock<Option<ProducerError>>>,
+        end_event: Arc<StickyEvent>,
+        flush_event: (Arc<EventHandler>, Arc<EventHandler>),
+    ) {
+        let producer =
+            PartitionProducer::shared(replica, spu_pool, batches, batch_events, linger, error);
+        fluvio_future::task::spawn(async move {
+            producer.run(end_event, flush_event).await;
+        });
+    }
+
+    #[instrument(skip(self, end_event, flush_event))]
+    async fn run(
+        &self,
+        end_event: Arc<StickyEvent>,
+        flush_event: (Arc<EventHandler>, Arc<EventHandler>),
+    ) {
+        use tokio::select;
+
+        let mut linger_sleep = sleep(std::time::Duration::from_secs(1800));
```

Comment:
lingering time should be getting from config value

Reply:
I am just adding a large value as linger time by default, and this is only changed to the config value once we get a new batch event

Reply:
I will change this to None, so it is more clear